### PR TITLE
MULE-19476: Refactor creation of reactor chain for processing strateg…

### PR DESCRIPTION
**Pull request by fsgonz**

### CODE REVIEW

File: core-tests/src/test/java/org/mule/runtime/core/internal/processor/strategy/processor/ProactorProcessingStrategyEnricherTestCase.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  @Spy
+  private ImmediateScheduler dispatcherScheduler;
+  @Mock
+  private CoreEvent coreEvent;
+
+  public ProactorProcessingStrategyEnricherTestCase(int parallelism) {
+    this.parallelism = parallelism;
+  }
+
+  @Parameters(name = "{0}")
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/internal/processor/strategy/processor/ProactorProcessingStrategyEnricherTestCase.java:60]: `"{0}"` -> `parallelism: "{0}"`

File: core-tests/src/test/java/org/mule/runtime/core/internal/processor/strategy/processor/ProcessingTypeBasedReactiveProcessorEnricherTestCase.java [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+  private ImmediateScheduler cpuLiteAsyncScheduler;
+
+  @Mock
+  private CoreEvent coreEvent;
+
+  public ProcessingTypeBasedReactiveProcessorEnricherTestCase(int parallelism) {
+    this.parallelism = parallelism;
+  }
+
+  @Parameterized.Parameters(name = "{0}")
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/internal/processor/strategy/processor/ProcessingTypeBasedReactiveProcessorEnricherTestCase.java:69]: `"{0}"` -> `parallelism: "{0}"`

File: core-tests/src/test/java/org/mule/runtime/core/internal/processor/strategy/reactor/builder/ComponentProcessingStrategyReactiveProcessorBuilderTestCase.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  @Mock
+  private CoreEvent coreEvent;
+
+  public ComponentProcessingStrategyReactiveProcessorBuilderTestCase(int parallelism) {
+    this.parallelism = parallelism;
+  }
+
+  @Parameters(name = "{0}")
+  public static Collection<Integer> paralellism() {
+    return asList(1, 4);
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/internal/processor/strategy/reactor/builder/ComponentProcessingStrategyReactiveProcessorBuilderTestCase.java:71]: `"{0}"` -> `parallelism: "{0}"`

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/enricher/CpuLiteAsyncNonBlockingProcessingStrategyEnricher.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+package org.mule.runtime.core.internal.processor.strategy.enricher;
+
+import org.mule.runtime.api.scheduler.Scheduler;
+import org.mule.runtime.core.api.processor.ReactiveProcessor;
+
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.function.Supplier;
+
+import static org.mule.runtime.core.internal.processor.strategy.reactor.builder.ComponentProcessingStrategyReactiveProcessorBuilder.processingStrategyReactiveProcessorFrom;
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/processor/strategy/enricher/CpuLiteAsyncNonBlockingProcessingStrategyEnricher.java:16]: static imports first

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/enricher/CpuLiteNonBlockingProcessingStrategyEnricher.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+ * LICENSE.txt file.
+ */
+
+package org.mule.runtime.core.internal.processor.strategy.enricher;
+
+import org.mule.runtime.api.scheduler.Scheduler;
+import org.mule.runtime.core.api.processor.ReactiveProcessor;
+import java.util.function.Supplier;
+
+import static org.mule.runtime.core.internal.processor.strategy.reactor.builder.ComponentProcessingStrategyReactiveProcessorBuilder.processingStrategyReactiveProcessorFrom;
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/processor/strategy/enricher/CpuLiteNonBlockingProcessingStrategyEnricher.java:14]: static imports first

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/enricher/ProactorProcessingStrategyEnricher.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        .withParallelism(getChainParallelism(processor))
+        .build();
+  }
+
+  private int getChainParallelism(ReactiveProcessor processor) {
+    if (maxConcurrency == 1) {
+      return 1;
+    } else if (maxConcurrency == MAX_VALUE) {
+      if (processor instanceof ComponentInnerProcessor && !((ComponentInnerProcessor) processor).isBlocking()) {
+        // For a no concurrency limit non blocking processor, the java SDK already handles parallelism internally, so no need to
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/processor/strategy/enricher/ProactorProcessingStrategyEnricher.java:60]: de we want to create a techDebt issue/TODO to tackle this eventually?

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/reactor/builder/ComponentProcessingStrategyReactiveProcessorBuilder.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+/**
+ * Builder for a {@link ReactiveProcessor} that enriches a component {@link ReactiveProcessor} with processing strategy logic. The
+ * processing strategy involves two decisions:
+ * <p>
+ * If the component task should be submitted using a {@link Scheduler} or if it should be done in the same thread.
+ * <p>
+ * If the response should be submitted using a {@link Scheduler} for returning to the flow or further processing can be done in
+ * the same thread.
+ * <p>
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/processor/strategy/reactor/builder/ComponentProcessingStrategyReactiveProcessorBuilder.java:31]: remove extra separator

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/reactor/builder/PipelineProcessingStrategyReactiveProcessorBuilder.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+import java.util.Optional;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.function.Function;
+
+/**
+ * Builder for a {@link ReactiveProcessor} that enriches a pipeline {@link ReactiveProcessor} with processing strategy logic. The
+ * processing strategy involves a decision regarding
+ * <p>
+ * the {@link Scheduler} to use for dispatching the a {@link CoreEvent} to a flow.
+ * <p>
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/processor/strategy/reactor/builder/PipelineProcessingStrategyReactiveProcessorBuilder.java:30]: remove extra

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/reactor/builder/PipelineProcessingStrategyReactiveProcessorBuilder.java [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+ *
+ * @since 4.4.0
+ */
+public class PipelineProcessingStrategyReactiveProcessorBuilder {
+
+  private final ReactiveProcessor pipeline;
+  private final ClassLoader executionClassloader;
+  private Optional<ScheduledExecutorService> scheduler = empty();
+  private Function<ScheduledExecutorService, ScheduledExecutorService> schedulerDecorator = identity();
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/processor/strategy/reactor/builder/PipelineProcessingStrategyReactiveProcessorBuilder.java:39]: you may use `UnaryOperator` here

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/reactor/builder/ReactorPublisherBuilder.java [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+import org.reactivestreams.Subscription;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.util.context.Context;
+
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import static reactor.core.scheduler.Schedulers.fromExecutorService;
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/processor/strategy/reactor/builder/ReactorPublisherBuilder.java:22]: static imports first

**Review by elrodro83 — CHANGES_REQUESTED** [review]

**Review by IvanAndresFritzler — CHANGES_REQUESTED** [review]

**Review by IvanAndresFritzler — APPROVED** [review]

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/enricher/ProactorProcessingStrategyEnricher.java (hunk 6 above)
Comment (fsgonz) [reply core/src/main/java/org/mule/runtime/core/internal/processor/strategy/enricher/ProactorProcessingStrategyEnricher.java:61]: I've created MULE-19526

**Review by elrodro83 — APPROVED** [review]

**Pull request merged**